### ISSUE TITLE
How to build for M1 macs?

### ISSUE THREAD

**Issue by weisJ** [nokeedev/gradle-native#294]
Is it possible to add new `MachineArchitectures`? As far as I can tell it should be possible to compile for M1 from intel macs using the `--target=arm64-apple-macos` option.

**Comment by weisJ** [comment]
I tried to hijack the x86 configuration to determine what the compiler/linker flags need to be.
So far the following needs to be added:
`sdkPath` is the output of`xcrun --sdk $version --show-sdk-path`.

Compiler:
- `-target arm64-apple-macos11`
- `-isysroot sdkPath`
- `-isystem sdkPath/usr/include` it defaults to the `10.15` include path for me, so I have to overwrite it explicitly.
Linker:
- `-target arm64-apple-macos11`
- `-isysroot sdkPath`
- `-isystem sdkPath/usr/include`

However this setup still is incompatible with `JavaNativeFoundation`:
````
ld: warning: ignoring file /Users/ulrike/.gradle/caches/transforms-3/7a3141fa7d07c4ec3342ec99b75028ac/transformed/JavaNativeFoundation.framework/JavaNativeFoundation.tbd, missing required architecture arm64 in file /Users/ulrike/.gradle/caches/transforms-3/7a3141fa7d07c4ec3342ec99b75028ac/transformed/JavaNativeFoundation.framework/JavaNativeFoundation.tbd
Undefined symbols for architecture arm64:
  "_JNFNativeMethodEnter", referenced from:
 ...
````

**Comment by weisJ** [comment]
This seems relevant:
https://developer.apple.com/forums/thread/654601

**Comment by lacasseio** [comment]
I recently (last week) removed a limitation regarding configuring target machine for any OS/Arch combination. This allows for same host builds without issue. Cross-compilation has a limitation around toolchain selection that I think we can get rid-of to allow this use case (and most likely all other use cases like Android and embedded compilation). It will require some manual configuration of the toolchains. Let me give it a try.

For `JavaNativeFoundation` pre-compiled ARM, it should be possible to provide it. You would have to configure dependencies per variant and switch the coordinate for ARM to the pre-built one. You can exposed it as another Gradle project or, if you publish it in a Maven repository, then you can resolve it. However, for both ways I would need to try it out first to see if there is any change that needs to be done. There are some opinion that we would need to iron out to make the integration smooth. What would be your preferred choice? As a side note, make sure the licensing allows you to build and publish the library on Maven. I had an issue once with STM toolchains that don't allow for "redistribution" even if it was only for my own project. I had placed the toolchain in a more convenient location for Gradle to provision them and that got a copyright violation (or whatever is the equivalent for GitHub).

**Comment by weisJ** [comment]
That sounds great. The apple openjdk port is licensed under GPL 3, so publishing compiled binaries should be fine. I have created a separate repository to publish it: https://github.com/weisJ/JavaNativeFoundation
I am now able to resolve the dependency locally but it definitely isn't configured correctly.
This is the error I am getting:
````
No matching variant of com.github.weisj:java-native-foundation:1.0.0 was found. The consumer was configured to find a usage of 'native-link' of a component, as well as attribute 'org.gradle.native.operatingSystem' with value 'macos', attribute 'org.gradle.native.architecture' with value 'x86-64', attribute 'org.gradle.native.debuggable' with value 'true', attribute 'org.gradle.native.optimized' with value 'false' but:
````
Could you give me a hint for what I need to setup such that the note plugin will resolve the framework correctly?

**Comment by lacasseio** [comment]
That is the error I though you would encounter... Since I'm working in the runtime code at the moment, I will polish the framework dependency for macOS. Hopefully I can have something tomorrow.

**Comment by lacasseio** [comment]
For the toolchain, you could give the following Software Model rule and build script snippet a try. It works by moving the complexity into toolchain definition which is where we want to go in the long run. We have to reprobe the toolchain a Gradle probing logic isn't great when it comes to custom `sysroot`. What we do is reprobe the tools and use a last chance argument handler to replace wrong probing result with the correct one. For the custom architecture to work, you will need to use one of the latest version of Nokee, [like version `0.4.0-202106241527.dd7a787e`](https://services.nokee.dev/versions/latest-snapshot.json). Hopefully there won't be any issue with upgrading to that version. If there is any we will work together in solving them.

#### `build.gradle`
```
apply plugin: dev.nokee.samples.AppleM1ToolChainRule

library {
	targetMachines = [machines.macOS.architecture('arm64')]
}
```

#### `AppleM1ToolChainRule.java`
```
package dev.nokee.samples;

import org.gradle.api.Action;
import org.gradle.internal.os.OperatingSystem;
import org.gradle.internal.service.ServiceRegistry;
import org.gradle.model.Finalize;
import org.gradle.model.RuleSource;
import org.gradle.nativeplatform.toolchain.Clang;
import org.gradle.nativeplatform.toolchain.GccCommandLineToolConfiguration;
import org.gradle.nativeplatform.toolchain.GccPlatformToolChain;
import org.gradle.nativeplatform.toolchain.NativeToolChainRegistry;
import org.gradle.nativeplatform.toolchain.internal.gcc.DefaultGccPlatformToolChain;
import org.gradle.nativeplatform.toolchain.internal.gcc.metadata.GccMetadata;
import org.gradle.nativeplatform.toolchain.internal.gcc.metadata.GccMetadataProvider;
import org.gradle.nativeplatform.toolchain.internal.metadata.CompilerMetaDataProvider;
import org.gradle.process.ExecOperations;
import org.gradle.process.internal.ExecActionFactory;

import javax.annotation.Nullable;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.util.Arrays;
import java.util.List;
import java.util.ListIterator;
import java.util.stream.Stream;

import static org.gradle.nativeplatform.toolchain.internal.gcc.metadata.GccMetadataProvider.forClang;

@SuppressWarnings("UnstableApiUsage")
public class AppleM1ToolChainRule extends RuleSource {
    @Finalize
    public void configureToolchain(NativeToolChainRegistry toolChains, ServiceRegistry serviceRegistry) {
        SdkPathProvider sdkPath = new SdkPathProvider(serviceRegistry.get(ExecOperations.class));
        toolChains.create("m1clang", Clang.class, toolchain -> {
            toolchain.target("macosarm64", new ConfigurePlatformAction(serviceRegistry.get(ExecActionFactory.class), toolchain, sdkPath));
        });
    }

    private static final class SdkPathProvider {
        private final ExecOperations execOperations;
        @Nullable private String sdkPath = null;

        SdkPathProvider(ExecOperations execOperations) {
            this.execOperations = execOperations;
        }

        public String get() {
            if (sdkPath == null) {
                sdkPath = loadSdkPath();
            }
            return sdkPath;
        }

        private String loadSdkPath() {
            ByteArrayOutputStream outStream = new ByteArrayOutputStream();
            execOperations.exec(spec -> {
                spec.commandLine("xcrun", "--sdk", "macosx", "--show-sdk-path");
                spec.setStandardOutput(outStream);
            });
            return outStream.toString().trim();
        }
    }

    private static final class CompilerMetadataProvider {
        private final GccMetadataProvider clangMetadataProvider;
        private final Clang toolchain;
        private final GccCommandLineToolConfiguration toolConfiguration;
        private final SdkPathProvider sdkPath;
        @Nullable private GccMetadata metadata = null;

        CompilerMetadataProvider(GccMetadataProvider clangMetadataProvider, Clang toolchain, GccCommandLineToolConfiguration toolConfiguration, SdkPathProvider sdkPath) {
            this.clangMetadataProvider = clangMetadataProvider;
            this.toolchain = toolchain;
            this.toolConfiguration = toolConfiguration;
            this.sdkPath = sdkPath;
        }

        public GccMetadata get() {
            if (metadata == null) {
                metadata = loadMetadata();
            }
            return metadata;
        }

        private GccMetadata loadMetadata() {
            return clangMetadataProvider.getCompilerMetaData(toolchain.getPath(), this::configureCompiler).getComponent();
        }

        private void configureCompiler(CompilerMetaDataProvider.CompilerExecSpec spec) {
            spec.args(Arrays.asList("-target", "arm64-apple-macos11", "-isysroot", sdkPath.get()));

            // Search in the System path... or in the toolchain path
            if (toolchain.getPath().isEmpty()) {
                spec.executable(OperatingSystem.current().findInPath(toolConfiguration.getExecutable()));
            } else {
                spec.executable(toolchain.getPath().stream().map(this::createExecutableFile).filter(File::isFile).findFirst()
                        .orElseThrow(this::executableNotFoundException));
            }
        }

        private IllegalStateException executableNotFoundException() {
            return new IllegalStateException(String.format("Executable '%s' not found in %s.", toolConfiguration.getExecutable(), toolchain.getPath()));
        }

        private File createExecutableFile(File path) {
            return new File(path, toolConfiguration.getExecutable());
        }
    }

    private static final class ConfigureCompilerArgumentsAction implements Action<List<String>> {
        private final CompilerMetadataProvider compilerMetadata;
        private final SdkPathProvider sdkPath;

        public ConfigureCompilerArgumentsAction(CompilerMetadataProvider compilerMetadata, SdkPathProvider sdkPath) {
            this.compilerMetadata = compilerMetadata;
            this.sdkPath = sdkPath;
        }

        @Override
        public void execute(List<String> args) {
            ListIterator<String> iter = args.listIterator();
            while (iter.hasNext()) {
                String arg = iter.next();
                if (arg.equals("-isystem")) {
                    iter.remove(); // remove flag

                    // remove flag value
                    iter.next();
                    iter.remove();
                }
            }
            args.addAll(Arrays.asList("-target", "arm64-apple-macos11", "-isysroot", sdkPath.get()));
            compilerMetadata.get().getSystemLibraries().getIncludeDirs()
                    .stream().flatMap(it -> Stream.of("-isystem", it.getAbsolutePath())).forEach(args::add);
            args.addAll(Arrays.asList("-F", sdkPath.get() + "/System/Library/Frameworks"));
        }
    }

    private static final class ConfigureLinkerArgumentsAction implements Action<List<String>> {
        private final CompilerMetadataProvider linkerMetadata;
        private final SdkPathProvider sdkPath;

        public ConfigureLinkerArgumentsAction(CompilerMetadataProvider linkerMetadata, SdkPathProvider sdkPath) {
            this.linkerMetadata = linkerMetadata;
            this.sdkPath = sdkPath;
        }

        @Override
        public void execute(List<String> args) {
            args.addAll(Arrays.asList("-target", "arm64-apple-macos11", "-isysroot", sdkPath.get()));
            linkerMetadata.get().getSystemLibraries().getLibDirs().stream().flatMap(it -> Stream.of("-L", it.getAbsolutePath())).forEach(args::add);
            args.addAll(Arrays.asList("-F", sdkPath.get() + "/System/Library/Frameworks"));
        }
    }

    private static final class ConfigurePlatformAction implements Action<GccPlatformToolChain> {
        private final GccMetadataProvider metadataProvider;
        private final Clang toolchain;
        private final SdkPathProvider sdkPath;

        ConfigurePlatformAction(ExecActionFactory execActionFactory, Clang toolchain, SdkPathProvider sdkPath) {
            this.metadataProvider = forClang(execActionFactory);;
            this.toolchain = toolchain;
            this.sdkPath = sdkPath;
        }

        @Override
        public void execute(GccPlatformToolChain platform) {
            platform.getcCompiler().withArguments(forCCompiler(platform.getcCompiler()));
            platform.getCppCompiler().withArguments(forCppCompiler(platform.getCppCompiler()));
            platform.getLinker().withArguments(forLinker(platform.getLinker()));
            configureCompilerProbeArguments(platform);
        }

        private Action<List<String>> forCCompiler(GccCommandLineToolConfiguration cCompiler) {
            CompilerMetadataProvider cCompilerMetadata = new CompilerMetadataProvider(metadataProvider, toolchain, cCompiler, sdkPath);
            return new ConfigureCompilerArgumentsAction(cCompilerMetadata, sdkPath);
        }

        private Action<List<String>> forCppCompiler(GccCommandLineToolConfiguration cppCompiler) {
            CompilerMetadataProvider cppCompilerMetadata = new CompilerMetadataProvider(metadataProvider, toolchain, cppCompiler, sdkPath);
            return new ConfigureCompilerArgumentsAction(cppCompilerMetadata, sdkPath);
        }

        private Action<List<String>> forLinker(GccCommandLineToolConfiguration linker) {
            CompilerMetadataProvider linkerMetadata = new CompilerMetadataProvider(metadataProvider, toolchain, linker, sdkPath);
            return new ConfigureLinkerArgumentsAction(linkerMetadata, sdkPath);
        }

        private void configureCompilerProbeArguments(GccPlatformToolChain platform) {
            // Configure compiler probe args, not strictly needed but for completeness.
            ((DefaultGccPlatformToolChain) platform).getCompilerProbeArgs().clear();
            ((DefaultGccPlatformToolChain) platform).getCompilerProbeArgs().addAll(Arrays.asList("-target", "arm64-apple-macos11", "-isysroot", sdkPath.get()));
        }
    }
}
```

**Comment by weisJ** [comment]
Thank you very much! AppleM1ToolChainRule.java works as far as I can replace my manual logic with it and now it only fails to resolve the new JavaNativeFoundation (as it isn#t yet configured).
Though it looks like that 
````java
toolChains.create("m1clang", Clang.class, toolchain -> {
     toolchain.target("macosarm64", new ConfigurePlatformAction(serviceRegistry.get(ExecActionFactory.class), toolchain, sdkPath));
});
````
internally probes the installed Xcode version and tries to invoke `xcrun`. This causes the build to fail on windows/linux.

After looking at some of the sources and the errors I was getting I have added the following to my publication:
````kotlin
import dev.nokee.runtime.nativebase.internal.LibraryElements
import dev.nokee.runtime.nativebase.internal.ArtifactSerializationTypes
import dev.nokee.runtime.nativebase.MachineArchitecture as NokeeMachineArchitecture
import dev.nokee.runtime.nativebase.OperatingSystemFamily as NokeeOSFamily

val frameWorkConfig: Configuration by configurations.creating
frameWorkConfig.apply {
    attributes {
        // Nokee attributes
        attribute(
            LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE,
            objects.named(LibraryElements::class.java, LibraryElements.FRAMEWORK_BUNDLE)
        )
        attribute(
            ArtifactSerializationTypes.ARTIFACT_SERIALIZATION_TYPES_ATTRIBUTE,
            ArtifactSerializationTypes.DESERIALIZED
        )
        attribute(
            NokeeOSFamily.OPERATING_SYSTEM_ATTRIBUTE,
            objects.named(NokeeOSFamily::class.java, NokeeOSFamily.MACOS)
        )
        attribute(
            NokeeMachineArchitecture.ARCHITECTURE_ATTRIBUTE,
            NokeeMachineArchitecture.forName(NokeeMachineArchitecture.X86_64)
        )

        //Gradle attributes
        attribute(
            OperatingSystemFamily.OPERATING_SYSTEM_ATTRIBUTE,
            objects.named(OperatingSystemFamily::class.java, OperatingSystemFamily.MACOS)
        )
        attribute(
            MachineArchitecture.ARCHITECTURE_ATTRIBUTE,
            objects.named(MachineArchitecture::class.java, MachineArchitecture.X86_64)
        )
        attribute(CppBinary.OPTIMIZED_ATTRIBUTE, false)
        attribute(CppBinary.DEBUGGABLE_ATTRIBUTE, true)
    }
}

val frameworkComponent = components.findByName("native-link") as AdhocComponentWithVariants
frameworkComponent.addVariantsFromConfiguration(frameWorkConfig) {
    mapToMavenScope("compile")
}

publishing {
    repositories {
        ...
    }
    publications {
        create<MavenPublication>("maven") {
            groupId = project.group.toString()
            artifactId = project.name
            version = project.version.toString()
            from(components["native-link"])
            artifact(archiveJNF.get().outputs.files.singleFile) {
                builtBy(archiveJNF)
            }
        }
    }
}
````
I still isn't quite enough though. Note that this is just to try to get the library to replace the current `x86-64` build.
How do I have to configure the publication s.t. the headers get added as include paths to the compiler and the framework to the linker?

Trying to compile the `arm64` version doesn't quite get this far, as the local `dev.nokee.framework:*` frameworks don't declare that they are compatible with `arm64`:
````
> No matching variant of dev.nokee.framework:Cocoa:11.1 was found. The consumer was configured to find a usage of 'native-link' of a component, as well as attribute 'org.gradle.libraryelements' with value 'DefaultLibraryElements()', attribute 'dev.nokee.artifactSerializationType' with value 'deserialized', attribute 'dev.nokee.operatingSystem' with value 'macos', attribute 'org.gradle.native.operatingSystem' with value 'macos', attribute 'dev.nokee.architecture' with value 'arm64', attribute 'org.gradle.native.architecture' with value 'arm64', attribute 'org.gradle.native.debuggable' with value 'true', attribute 'org.gradle.native.optimized' with value 'false' but:
             - Variant 'compile' capability dev.nokee.framework:Cocoa:11.1 declares a component, as well as attribute 'org.gradle.libraryelements' with value 'framework-bundle':
                 - Incompatible because this component declares a usage of 'cplusplus-api' of a component and the consumer needed a usage of 'native-link' of a component
                 - Other compatible attributes:
                     - Doesn't say anything about dev.nokee.architecture (required 'arm64')
                     - Doesn't say anything about dev.nokee.artifactSerializationType (required 'deserialized')
                     - Doesn't say anything about dev.nokee.operatingSystem (required 'macos')
                     - Doesn't say anything about org.gradle.native.architecture (required 'arm64')
                     - Doesn't say anything about org.gradle.native.debuggable (required 'true')
                     - Doesn't say anything about org.gradle.native.operatingSystem (required 'macos')
                     - Doesn't say anything about org.gradle.native.optimized (required 'false')
             - Variant 'link' capability dev.nokee.framework:Cocoa:11.1 declares a usage of 'native-link' of a component, as well as attribute 'org.gradle.libraryelements' with value 'framework-bundle', attribute 'org.gradle.native.operatingSystem' with value 'macos':
                 - Incompatible because this component declares a component, as well as attribute 'dev.nokee.architecture' with value 'x86-64' and the consumer needed a component, as well as attribute 'dev.nokee.architecture' with value 'arm64'
                 - Other compatible attributes:
                     - Doesn't say anything about dev.nokee.artifactSerializationType (required 'deserialized')
                     - Doesn't say anything about dev.nokee.operatingSystem (required 'macos')
                     - Doesn't say anything about org.gradle.native.architecture (required 'arm64')
                     - Doesn't say anything about org.gradle.native.debuggable (required 'true')
                     - Doesn't say anything about org.gradle.native.optimized (required 'false')
             - Variant 'runtime' capability dev.nokee.framework:Cocoa:11.1 declares a component, as well as attribute 'org.gradle.native.operatingSystem' with value 'macos':
                 - Incompatible because this component declares a usage of 'native-runtime' of a component, as well as attribute 'dev.nokee.architecture' with value 'x86-64' and the consumer needed a usage of 'native-link' of a component, as well as attribute 'dev.nokee.architecture' with value 'arm64'
                 - Other compatible attributes:
                     - Doesn't say anything about dev.nokee.artifactSerializationType (required 'deserialized')
                     - Doesn't say anything about dev.nokee.operatingSystem (required 'macos')
                     - Doesn't say anything about org.gradle.libraryelements (required 'DefaultLibraryElements()')
                     - Doesn't say anything about org.gradle.native.architecture (required 'arm64')
                     - Doesn't say anything about org.gradle.native.debuggable (required 'true')
                     - Doesn't say anything about org.gradle.native.optimized (required 'false')
````

**Comment by lacasseio** [comment]
The "local" framework that Nokee support is highly opinionated. Given it was never suppose to resolve from a 3rd party repository, it was good enough. In your case, you would want to resolve a published framework so I need to come up with something that is stable enough to ensure we can grow. I'm looking right now at the right dependency rules/attributes to set on the artifact so it resolves properly. Hopefully later today I can have a Nokee version that works.

Regarding Apple M1 toolchain, I think we would just have to avoid querying `xcrun` if the tool is not available returning a dummy sdk path in those cases. We would need to change a bit the implementation for `SdkPathProvider`.

**Comment by weisJ** [comment]
> Regarding Apple M1 toolchain, I think we would just have to avoid querying xcrun if the tool is not available returning a dummy sdk path in those cases. We would need to change a bit the implementation for SdkPathProvider.

The problem isn't actually the `SdkPathProvider`. Even
````kotlin
toolChains.create("m1clang", Clang.class, toolchain -> {
     toolchain.target("macosarm64");
});
````
alone causes the following error:
````
Build darklaf FAILURE reason:                                
    Caused by: org.gradle.api.ProjectConfigurationException: A problem occurred configuring project ':darklaf-macos'.
        at org.gradle.configuration.project.LifecycleProjectEvaluator.wrapException(LifecycleProjectEvaluator.java:75)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.addConfigurationFailure(LifecycleProjectEvaluator.java:68)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.access$400(LifecycleProjectEvaluator.java:51)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate.run(LifecycleProjectEvaluator.java:191)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.lambda$run$0(LifecycleProjectEvaluator.java:105)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$applyToMutableState$0(DefaultProjectStateRegistry.java:250)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.lambda$withProjectLock$3(DefaultProjectStateRegistry.java:310)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:213)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.withProjectLock(DefaultProjectStateRegistry.java:310)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.fromMutableState(DefaultProjectStateRegistry.java:291)
        at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.applyToMutableState(DefaultProjectStateRegistry.java:249)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.run(LifecycleProjectEvaluator.java:91)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.evaluate(LifecycleProjectEvaluator.java:63)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:723)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:150)
        at org.gradle.execution.TaskPathProjectEvaluator.configure(TaskPathProjectEvaluator.java:41)
        at org.gradle.execution.TaskPathProjectEvaluator.configureHierarchy(TaskPathProjectEvaluator.java:69)
        at org.gradle.configuration.DefaultProjectsPreparer.prepareProjects(DefaultProjectsPreparer.java:46)
        at org.gradle.configuration.BuildTreePreparingProjectsPreparer.prepareProjects(BuildTreePreparingProjectsPreparer.java:64)
        at org.gradle.configuration.BuildOperationFiringProjectsPreparer$ConfigureBuild.run(BuildOperationFiringProjectsPreparer.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
        at org.gradle.configuration.BuildOperationFiringProjectsPreparer.prepareProjects(BuildOperationFiringProjectsPreparer.java:40)
        at org.gradle.initialization.DefaultGradleLauncher.prepareProjects(DefaultGradleLauncher.java:226)
        at org.gradle.initialization.DefaultGradleLauncher.doClassicBuildStages(DefaultGradleLauncher.java:164)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:149)
        at org.gradle.initialization.DefaultGradleLauncher.executeTasks(DefaultGradleLauncher.java:125)
        org.gradle.internal.event.ListenerNotificationException: Failed to notify project evaluation listener.
            at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:89)
            at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:346)
            at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:249)
            at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:141)
            at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
            at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
            at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate$1.execute(LifecycleProjectEvaluator.java:183)
            at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate$1.execute(LifecycleProjectEvaluator.java:180)
            at org.gradle.api.internal.project.DefaultProject.stepEvaluationListener(DefaultProject.java:1454)
            at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate.run(LifecycleProjectEvaluator.java:189)
            ... 51 more
            Cause 1: org.gradle.process.internal.ExecException: A problem occurred starting process 'command 'xcrun''
                at org.gradle.process.internal.DefaultExecHandle.execExceptionFor(DefaultExecHandle.java:241)
                at org.gradle.process.internal.DefaultExecHandle.setEndStateInfo(DefaultExecHandle.java:218)
                at org.gradle.process.internal.DefaultExecHandle.failed(DefaultExecHandle.java:369)
                at org.gradle.process.internal.ExecHandleRunner.run(ExecHandleRunner.java:87)
                at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:42)
                at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
                at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
                at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
                Caused by: net.rubygrapefruit.platform.NativeException: Could not start 'xcrun'
                    at net.rubygrapefruit.platform.internal.DefaultProcessLauncher.start(DefaultProcessLauncher.java:27)
                    at net.rubygrapefruit.platform.internal.WindowsProcessLauncher.start(WindowsProcessLauncher.java:22)
                    at net.rubygrapefruit.platform.internal.WrapperProcessLauncher.start(WrapperProcessLauncher.java:36)
                    at org.gradle.process.internal.ExecHandleRunner.startProcess(ExecHandleRunner.java:98)
                    at org.gradle.process.internal.ExecHandleRunner.run(ExecHandleRunner.java:71)
                    ... 4 more
                    Caused by: java.io.IOException: Cannot run program "xcrun" (in directory "C:\Users\weisj\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\211.6693.111\jbr\bin"): CreateProcess error=2, The system cannot find the file specified
                        at net.rubygrapefruit.platform.internal.DefaultProcessLauncher.start(DefaultProcessLauncher.java:25)
                        ... 8 more
                        Caused by: java.io.IOException: CreateProcess error=2, The system cannot find the file specified
                            ... 9 more
            Cause 2: org.gradle.process.internal.ExecException: A problem occurred starting process 'command 'xcrun''
                at org.gradle.process.internal.DefaultExecHandle.execExceptionFor(DefaultExecHandle.java:241)
                at org.gradle.process.internal.DefaultExecHandle.setEndStateInfo(DefaultExecHandle.java:218)
                at org.gradle.process.internal.DefaultExecHandle.failed(DefaultExecHandle.java:369)
                at org.gradle.process.internal.ExecHandleRunner.run(ExecHandleRunner.java:87)
                at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:42)
                at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
                at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
                at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
                at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
                at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
                at java.base/java.lang.Thread.run(Thread.java:834)
                Caused by: net.rubygrapefruit.platform.NativeException: Could not start 'xcrun'
                    at net.rubygrapefruit.platform.internal.DefaultProcessLauncher.start(DefaultProcessLauncher.java:27)
                    at net.rubygrapefruit.platform.internal.WindowsProcessLauncher.start(WindowsProcessLauncher.java:22)
                    at net.rubygrapefruit.platform.internal.WrapperProcessLauncher.start(WrapperProcessLauncher.java:36)
                    at org.gradle.process.internal.ExecHandleRunner.startProcess(ExecHandleRunner.java:98)
                    at org.gradle.process.internal.ExecHandleRunner.run(ExecHandleRunner.java:71)
                    ... 7 more
                    Caused by: java.io.IOException: Cannot run program "xcrun" (in directory "C:\Users\weisj\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\211.6693.111\jbr\bin"): CreateProcess error=2, The system cannot find the file specified
                        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1128)
                        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1071)
                        at net.rubygrapefruit.platform.internal.DefaultProcessLauncher.start(DefaultProcessLauncher.java:25)
                        ... 11 more
                        Caused by: java.io.IOException: CreateProcess error=2, The system cannot find the file specified
                            at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:478)
                            at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:154)
                            at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1107)
                            ... 13 more


FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':darklaf-macos'.
> Failed to notify project evaluation listener.
   > A problem occurred starting process 'command 'xcrun''
   > A problem occurred starting process 'command 'xcrun''
````
Though it only happens when a target with `arm64` is actually declared.

**Comment by lacasseio** [comment]
Ah, I think I know what's happening. The call to `target("macosarm64")` only append another target to the default target. You will have to make a call to overwrite the targets: `setTargets("macosarm64")` (https://github.com/gradle/gradle/blob/29488b556746fe6e344532ea12a2190aed40e83a/subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/GccCompatibleToolChain.java#L57). Then call to configure the target. If that doesn't work, I will try it on my Windows VM to debug further. My guess is the `xcrun` comes from the compiler probing in Gradle and Nokee is trying to select a toolchain. We may be trying too early to select the toolchain, however, the M1 toolchain should return as unavailable for the selected platform (macOS on ARM64) so we should not proceed further.

**Comment by weisJ** [comment]
This doesn't quite seem to work. However removing `machines.macOS.architecture("arm64")` from the defined `targetMachines` resolve the issues.

**Comment by lacasseio** [comment]
Thanks for trying it out. I will test it on a Windows VM and figure out what is happening. I haven't forgotten about the 3rd party Apple Framework, I'm simply trying to come up with the right attribute declaration to support adhoc framework, subproject framework and external (Maven) framework. I have something working but need to add more tests and change a little bit how dependencies are consumed internally in Nokee. That will open up the possibility to consume framework instead off includes directory/shared library regardless of where the framework comes from.

**Comment by lacasseio** [comment]
I can confirm the failure on Windows is due to how toolchain selection works in Gradle. It's annoying but I will find a way to work around it.

I'm working on [this PR](https://github.com/nokeedev/gradle-native/pull/300) to allow external framework. I also have a patch to your JavaNativeFoundation recompile for M1 which seems to be work for JNI. There are some small validation that I still need to make so hopefully I can finish all this tomorrow.

**Comment by lacasseio** [comment]
I opened [this PR](https://github.com/weisJ/JavaNativeFoundation/pull/1) on your JavaNativeFoundation repo. I used a Groovy DSL script to create the adhoc plugin simply because I'm not skilled with Kotlin, feel free to convert it back if you want.

I'm waiting for the test to pass on my PR which also solve/workaround the `xcrun` execution on Windows issue. It's not the final solution but it's good enough to unblock you. I published version `0.4.231-202106301216.6aa4b5ee` based on the latest changes of the PR. As tests haven't finish running, I'm unsure if there is any failures/breakage. Feel free to give it a try and report back.

**Comment by weisJ** [comment]
Thank you very much for your efforts! With the adjusted JavaNativeFoundation artifact I was able to replace the current `x86-64` target variant. However I still can't yet enable the `arm64` target due to them not declaring that they are compatible with `arm64`. Also `xcrun` still causes an issue on Windows for me :/

**Comment by weisJ** [comment]
I stand to be corrected. `xcrun` actually no longer fails the build on Windows (I think I was running an older cached version of nokee). However `SharedLibraryBinary#isBuildable` now throws an error. For now I simply wrapped the call in a try catch block.

**Comment by lacasseio** [comment]
Ah yes, you are correct. I catch the exception in the toolchain selection but use the underlying Gradle API in the binary. I will catch the exception for now until I rework this. This change was a bit bigger than expected but I think it will unblock quite a few use cases. It's especially useful for M1 compilation which was on my list since January. :-)

**Comment by lacasseio** [comment]
This took longer than expected but it really helped push some good story forward. I merged the PR and cut a release. Give version `0.4.252-202107041328.d6bc685f` a try.

**Comment by weisJ** [comment]
I am still getting an `xcrun` exception when calling `SharedLibraryBinary#isBuildable` on windows. Also the `dev.nokee.framework:AppKit` and `dev.nokee.framework:Cocoa` frameworks don't seem to offer an `arm64` version. Notably this is when trying to cross compile from an intel mac (I don't have access to an M1 mac, so I can't test compilation on there).

**Comment by lacasseio** [comment]
Thanks for the feedback. I will have a look at these issues tomorrow. For the `isBuildable` exception, I must have failed with code searching.

For `AppKit` and `Cocoa` frameworks, we don't search for other targets so I will have to include that in so we find them. In theory, if you know where they are located, you could use subprojects with outgoing configuration configured like [this](https://github.com/nokeedev/gradle-native/blob/master/subprojects/runtime-darwin/src/functionalTest/groovy/dev/nokee/runtime/darwin/OutgoingElements.groovy#L26-L33). To be correct, you should also set the artifact type to `framework` on the outgoing artifact. Unfortunately, you can't hack up direct file dependencies to the framework because of https://github.com/nokeedev/gradle-native/issues/301. I only recommend this as a quick solution to unblock you so we can see if there is any issue further down the compilation.

**Comment by weisJ** [comment]
Actually the `sdk` path for version `11` (and later) which has to be used when building for `arm64` contains `tbd` definition files for which support both `x86-64` and `arm64`. By manually specifying the framework linker flags for `AppKit` and `Cocoa` I am now able to build the library.

**Comment by lacasseio** [comment]
PR https://github.com/nokeedev/gradle-native/pull/303 should fix the `isBuildable()` exception. PR https://github.com/nokeedev/gradle-native/pull/304 is still a work in progress and will honours targets for Framework.

**Comment by lacasseio** [comment]
I made a modification to the framework discovery to account for the target information inside the TBD file. Give version `0.4.263-202107071002.6b1cdad5` a try.

**Comment by lacasseio** [comment]
There was a failure in a Framework dependency sample, so I cut another release with the fix: `0.4.264-202107071245.986a5b8a`

**Comment by weisJ** [comment]
The new version it no longer used the version >= 11 framework for arm61 (it did before) but falls back to version 10.15.*, which, using the m1-toolchain rule, actually won't be found.

**Comment by lacasseio** [comment]
That is strange, I was able to see a version 11 get picked up in my tests. I will give it a try with your project and debug from there.

**Comment by weisJ** [comment]
Ok so I was able to test this on a non GitHub actions runner and there it works fine. This may be some idiosyncrasy with the action runner setup.

**Comment by weisJ** [comment]
Update: Sometimes the local framework resolution times out (which is quite weird as it is a local path):
````
Caused by: org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$ArtifactResolveException: Could not resolve all artifacts for configuration ':darklaf-macos:arm64NativeLinkLibraries'.
                at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.rethrowFailure(DefaultConfiguration.java:1420)
                at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.access$3600(DefaultConfiguration.java:150)
                at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$DefaultResolutionHost.rethrowFailure(DefaultConfiguration.java:2032)
                at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationArtifactCollection.lambda$new$0(DefaultConfiguration.java:1976)
                at org.gradle.internal.model.CalculatedValueContainerFactory$SupplierBackedCalculator.calculateValue(CalculatedValueContainerFactory.java:78)
                at org.gradle.internal.model.CalculatedValueContainer$CalculationState.lambda$attachValue$0(CalculatedValueContainer.java:216)
                at org.gradle.internal.Try.ofFailable(Try.java:39)
                at org.gradle.internal.model.CalculatedValueContainer$CalculationState.attachValue(CalculatedValueContainer.java:211)
                at org.gradle.internal.model.CalculatedValueContainer.finalizeIfNotAlready(CalculatedValueContainer.java:184)
                at org.gradle.internal.model.CalculatedValueContainer.finalizeIfNotAlready(CalculatedValueContainer.java:175)
                at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationArtifactCollection.ensureResolved(DefaultConfiguration.java:2012)
                at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationArtifactCollection.getArtifacts(DefaultConfiguration.java:1989)
                at dev.nokee.platform.nativebase.internal.dependencies.DefaultNativeIncomingDependencies.lambda$fromLinkConfiguration$1(DefaultNativeIncomingDependencies.java:94)
                at org.gradle.api.internal.provider.DefaultProvider.calculateOwnValue(DefaultProvider.java:66)
                at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateValue(AbstractMinimalProvider.java:103)
                at org.gradle.api.internal.provider.Collectors$ElementsFromCollectionProvider.collectEntries(Collectors.java:216)
                at org.gradle.api.internal.provider.AbstractCollectionProperty$CollectingSupplier.calculateValue(AbstractCollectionProperty.java:337)
                at org.gradle.api.internal.provider.AbstractCollectionProperty.finalValue(AbstractCollectionProperty.java:189)
                at org.gradle.api.internal.provider.AbstractCollectionProperty.finalValue(AbstractCollectionProperty.java:37)
                at org.gradle.api.internal.provider.AbstractProperty.finalizeNow(AbstractProperty.java:236)
                ... 51 more
                Cause 1: org.gradle.api.internal.artifacts.transform.TransformException: Failed to transform AppKit.framework.localpath (dev.nokee.framework:AppKit:11.1) to match attributes {artifactType=framework, dev.nokee.architecture=arm64, dev.nokee.artifactSerializationType=deserialized, dev.nokee.operatingSystem=macos, org.gradle.libraryelements=framework-bundle, org.gradle.native.architecture=arm64, org.gradle.native.operatingSystem=macos, org.gradle.status=release, org.gradle.usage=cplusplus-api+native-link}.
                    at org.gradle.api.internal.artifacts.transform.TransformingAsyncArtifactListener$TransformedArtifact.lambda$visit$2(TransformingAsyncArtifactListener.java:221)
                    at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:268)
                    at org.gradle.api.internal.artifacts.transform.TransformingAsyncArtifactListener$TransformedArtifact.visit(TransformingAsyncArtifactListener.java:213)
                    at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ParallelResolveArtifactSet$VisitingSet$StartVisitAction.visitResults(ParallelResolveArtifactSet.java:99)
                    at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ParallelResolveArtifactSet$VisitingSet.visit(ParallelResolveArtifactSet.java:68)
                    at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.visitArtifacts(DefaultLenientConfiguration.java:263)
                    at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.access$500(DefaultLenientConfiguration.java:72)
                    at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$2.run(DefaultLenientConfiguration.java:237)
                    at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
                    at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
                    at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
                    at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
                    at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
                    at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
                    at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
                    at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
                    at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
                    at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
                    at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.visitArtifactsWithBuildOperation(DefaultLenientConfiguration.java:234)
                    at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.access$200(DefaultLenientConfiguration.java:72)
                    at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$1.visitArtifacts(DefaultLenientConfiguration.java:137)
                    at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationArtifactCollection.lambda$new$0(DefaultConfiguration.java:1970)
                    ... 67 more
                    Caused by: org.gradle.internal.resolve.ArtifactResolveException: Could not download AppKit.framework.localpath (dev.nokee.framework:AppKit:11.1)
                        at org.gradle.api.internal.artifacts.repositories.resolver.ExternalResourceResolver$RemoteRepositoryAccess.resolveArtifact(ExternalResourceResolver.java:505)
                        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.CachingModuleComponentRepository$ResolveAndCacheRepositoryAccess.resolveArtifact(CachingModuleComponentRepository.java:462)
                        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.ErrorHandlingModuleComponentRepository$ErrorHandlingModuleComponentRepositoryAccess.lambda$resolveArtifact$15(ErrorHandlingModuleComponentRepository.java:181)
                        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.ErrorHandlingModuleComponentRepository$ErrorHandlingModuleComponentRepositoryAccess.tryResolveAndMaybeBlacklist(ErrorHandlingModuleComponentRepository.java:237)
                        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.ErrorHandlingModuleComponentRepository$ErrorHandlingModuleComponentRepositoryAccess.performOperationWithRetries(ErrorHandlingModuleComponentRepository.java:203)
                        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.ErrorHandlingModuleComponentRepository$ErrorHandlingModuleComponentRepositoryAccess.resolveArtifact(ErrorHandlingModuleComponentRepository.java:179)
                        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.FilteredModuleComponentRepository$FilteringAccess.resolveArtifact(FilteredModuleComponentRepository.java:142)
                        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.RepositoryChainArtifactResolver.resolveArtifact(RepositoryChainArtifactResolver.java:102)
                        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.DefaultArtifactSet$LazyArtifactSupplier.calculateValue(DefaultArtifactSet.java:273)
                        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.DefaultArtifactSet$LazyArtifactSupplier.calculateValue(DefaultArtifactSet.java:244)
                        at org.gradle.internal.model.CalculatedValueContainer$CalculationState.lambda$attachValue$0(CalculatedValueContainer.java:216)
                        at org.gradle.internal.Try.ofFailable(Try.java:39)
                        at org.gradle.internal.model.CalculatedValueContainer$CalculationState.attachValue(CalculatedValueContainer.java:211)
                        at org.gradle.internal.model.CalculatedValueContainer.finalizeIfNotAlready(CalculatedValueContainer.java:184)
                        at org.gradle.internal.model.CalculatedValueContainer.finalizeIfNotAlready(CalculatedValueContainer.java:175)
                        at org.gradle.api.internal.artifacts.transform.TransformingAsyncArtifactListener$TransformedArtifact.finalizeValue(TransformingAsyncArtifactListener.java:176)
                        at org.gradle.api.internal.artifacts.transform.TransformingAsyncArtifactListener$TransformedArtifact.run(TransformingAsyncArtifactListener.java:125)
                        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
                        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
                        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
                        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
                        at org.gradle.internal.operations.DefaultBuildOperationExecutor$QueueWorker.execute(DefaultBuildOperationExecutor.java:199)
                        at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.runOperation(DefaultBuildOperationQueue.java:241)
                        at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.doRunBatch(DefaultBuildOperationQueue.java:222)
                        at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.lambda$runBatch$0(DefaultBuildOperationQueue.java:212)
                        at org.gradle.internal.resources.AbstractResourceLockRegistry.whileDisallowingLockChanges(AbstractResourceLockRegistry.java:70)
                        at org.gradle.internal.work.DefaultWorkerLeaseService.whileDisallowingProjectLockChanges(DefaultWorkerLeaseService.java:185)
                        at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.lambda$runBatch$1(DefaultBuildOperationQueue.java:212)
                        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:213)
                        at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.runBatch(DefaultBuildOperationQueue.java:200)
                        at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.run(DefaultBuildOperationQueue.java:172)
                        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
                        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
                        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
                        Caused by: org.gradle.api.resources.ResourceException: Could not get resource 'http://127.0.0.1:49272/dev/nokee/framework/AppKit/11.1/AppKit.framework.localpath'.
                            at org.gradle.internal.resource.ResourceExceptions.failure(ResourceExceptions.java:74)
                            at org.gradle.internal.resource.ResourceExceptions.getFailed(ResourceExceptions.java:57)
                            at org.gradle.api.internal.artifacts.repositories.resolver.DefaultExternalResourceArtifactResolver.downloadByUrl(DefaultExternalResourceArtifactResolver.java:114)
                            at org.gradle.api.internal.artifacts.repositories.resolver.DefaultExternalResourceArtifactResolver.downloadStaticResource(DefaultExternalResourceArtifactResolver.java:92)
                            at org.gradle.api.internal.artifacts.repositories.resolver.DefaultExternalResourceArtifactResolver.resolveArtifact(DefaultExternalResourceArtifactResolver.java:62)
                            at org.gradle.api.internal.artifacts.repositories.resolver.ExternalResourceResolver$RemoteRepositoryAccess.resolveArtifact(ExternalResourceResolver.java:498)
                            ... 33 more
````

**Comment by lacasseio** [comment]
Sorry for dropping the ball on this issue and thanks for trying it out some more. I will keep working on this tomorrow.

**Comment by lacasseio** [comment]
I looked into the failures and unfortunately, I'm unable to see the same failure. My suggestion is to open a separate issue and post link to info logs for builds with the same failure. I know it's a bit troublesome but it will give me a sense of how often the issue appear and hopefully the logs would clue in on where the failure happens. The embedded server is something we want to get away from because of this exact flakiness but isn't schedule for work in the next 1-3 months. :-(

If you don't have any other issue with M1 compilation and remote framework resolution, we can close this issue. Thanks for trying out the fixes to unblock you.

**Comment by weisJ** [comment]
Thank you very much for all the effort. I figured that the timeout issue is hard to replicate. For now it doesn’t really bother me that much as it only happens rarely and probably is dependent on the load of the GitHub action runners. Once I have some logs to provide I’ll create a separate issue for the timeout. M1 Compilation works fine now, so I’ll close the issue :)
